include/boost/json/impl/serializer.ipp

```diff
@@ -784,8 +819,10 @@ serializer::
 reset(string const* p) noexcept
 {
     cs0_ = { p->data(), p->size() };
-    fn0_ = &serializer::write_string<true>;
-    fn1_ = &serializer::write_string<false>;
+    fn0_opt_ = &serializer::write_string<true>;
+    fn1_opt_ = &serializer::write_string<false>;
+    fn0_err_ = &serializer::write_string<true>;
+    fn1_err_ = &serializer::write_string<false>;
```

Comment:
hopefully this nonsense will go away when we make the options a data member

Reply:
This should make the code faster, since the compiler doesn't need to recheck which it is.

Reply:
To be fair, it's probably overdoing it, because this is one unnecessary if -else per double overhead.  I didn't profile either, just copied your model.